File: windows/Segmenter.Core/SegmentValidator.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace Segmenter.Core
{
    public class SegmentValidationError : Exception
    {
        public SegmentValidationError(string message) : base(message) { }
    }

    public class SegmentValidator
    {
        public const int MinDurationMs = 5_000;
        public const int MaxTimestampMs = 21_600_000;
        public const int MaxIntroDurationMs = 200_000;
        public const int MaxRecapDurationMs = 1_200_000;
        public const int MaxCreditsDurationMs = 1_800_000;
        public const int MaxPreviewDurationMs = 1_800_000;

        private static readonly Regex ImdbPattern = new Regex(@"^tt[0-9]{7,8}$");

        public static Dictionary<string, object?> MakeTheIntroDbSubmissionRequest(SubmissionDraft draft)
        {
            var imdbId = NormalizeImdb(draft.ImdbId);
            var hasTmdb = draft.TmdbId > 0;
            var hasImdb = imdbId != null && IsValidImdb(imdbId);

            if (!hasTmdb && !hasImdb)
            {
                throw new SegmentValidationError("At least TMDB ID or a valid IMDB ID must be provided");
            }

            if (draft.MediaType == MediaType.Movie)
            {
                if (draft.Season != null || draft.Episode != null)
                {
                    throw new SegmentValidationError("Season and episode must be empty for movie submissions");
                }
            }
            else if (draft.MediaType == MediaType.Tv)
            {
                if (draft.Season == null || draft.Season <= 0 || draft.Episode == null || draft.Episode <= 0)
                {
                    throw new SegmentValidationError("Season and episode are required for TV submissions");
                }
            }

            var validated = ValidateTheIntroDbRequest(draft);

            var payload = new Dictionary<string, object?>
            {
                { "type", ((MediaType)validated["type"]!).GetValue() },
                { "segment", ((SegmentType)validated["segment"]!).GetValue() },
                { "start_ms", validated["start_ms"] },
                { "tmdb_id", hasTmdb ? validated["tmdb_id"] : null }
            };

            if (validated.TryGetValue("season", out var s) && s != null)
                payload["season"] = s.ToString();
            if (validated.TryGetValue("episode", out var ep) && ep != null)
                payload["episode"] = ep.ToString();
            if (validated.TryGetValue("end_ms", out var end) && end != null)
                payload["end_ms"] = end;
            if (validated.TryGetValue("imdb_id", out var imdb) && imdb != null)
                payload["imdb_id"] = imdb;
            if (draft.VideoDurationMs != null)
                payload["video_duration_ms"] = draft.VideoDurationMs;

            return payload;
        }

        public static Dictionary<string, object?> MakeIntroDbSubmissionRequest(SubmissionDraft draft)
        {
            if (draft.MediaType != MediaType.Tv)
            {
                throw new SegmentValidationError("IntroDB supports TV episodes only");
            }

            var imdbId = NormalizeImdb(draft.ImdbId);
            if (string.IsNullOrEmpty(imdbId) || !IsValidImdb(imdbId))
            {
                throw new SegmentValidationError("Valid IMDB ID is required for IntroDB uploads");
            }

            if (draft.Season == null || draft.Season <= 0 || draft.Episode == null || draft.Episode <= 0)
            {
                throw new SegmentValidationError("Season and episode are required for IntroDB uploads");
            }

            var introDbSegment = ToIntroDbSegmentType(draft.Segment);
            if (string.IsNullOrEmpty(introDbSegment))
            {
                throw new SegmentValidationError($"IntroDB does not support {draft.Segment.DisplayName()} uploads");
            }

            var validated = ValidateTheIntroDbRequest(draft);
            if (!validated.TryGetValue("end_ms", out var endMs) || endMs == null)
            {
                throw new SegmentValidationError("IntroDB requires an explicit end timestamp");
            }

            return new Dictionary<string, object?>
            {
                { "segment_type", introDbSegment },
                { "imdb_id", imdbId },
                { "season", draft.Season },
                { "episode", draft.Episode },
                { "start_sec", Convert.ToDouble(validated["start_ms"]) / 1000.0 },
                { "end_sec", Convert.ToDouble(endMs) / 1000.0 },
                { "tvdb_id", null },
                { "tmdb_id", draft.TmdbId > 0 ? (int?)draft.TmdbId : null }
            };
        }

        private static string? ToIntroDbSegmentType(SegmentType segment)
        {
            return segment switch
            {
                SegmentType.Intro => "intro",
                SegmentType.Recap => "recap",
                SegmentType.Credits => "outro",
                _ => null
            };
        }

        private static Dictionary<string, object?> ValidateTheIntroDbRequest(SubmissionDraft draft)
        {
            return draft.Segment switch
            {
                SegmentType.Intro => ValidateIntroOrRecap(draft, MaxIntroDurationMs),
                SegmentType.Recap => ValidateIntroOrRecap(draft, MaxRecapDurationMs),
                SegmentType.Credits => ValidateCreditsOrPreview(draft, MaxCreditsDurationMs),
                SegmentType.Preview => ValidateCreditsOrPreview(draft, MaxPreviewDurationMs),
                _ => throw new SegmentValidationError("Unknown segment type")
            };
        }

        private static Dictionary<string, object?> ValidateIntroOrRecap(SubmissionDraft draft, int maxDurationMs)
        {
            var start = draft.StartMs;
            if (start == null)
            {
                if (draft.Segment == SegmentType.Intro)
                {
                    start = 0;
                }
                else
                {
                    throw new SegmentValidationError($"{draft.Segment.DisplayName()} start is required");
                }
            }
            else
            {
                start = Math.Max(start.Value, 0);
            }

            var end = draft.EndMs;
            if (end == null)
            {
                throw new SegmentValidationError($"{draft.Segment.DisplayName()} end is required");
            }

            AssertTimestampBounds(start.Value);
            AssertTimestampBounds(end.Value);

            if (end < start)
            {
                throw new SegmentValidationError("End must be greater than or equal to start");
            }

            var duration = end.Value - start.Value;
            if (duration != 0)
            {
                if (duration < MinDurationMs || duration > maxDurationMs)
                {
                    throw new SegmentValidationError(
                        $"{draft.Segment.DisplayName()} duration must be 0 or between " +
                        $"{MinDurationMs / 1000}s and {maxDurationMs / 1000}s"
                    );
                }
            }

            return new Dictionary<string, object?>
            {
                { "tmdb_id", draft.TmdbId },
                { "type", draft.MediaType },
                { "segment", draft.Segment },
                { "season", draft.Season },
                { "episode", draft.Episode },
                { "start_ms", start },
                { "end_ms", end },
                { "imdb_id", NormalizeImdb(draft.ImdbId) }
            };
        }

        private static Dictionary<string, object?> ValidateCreditsOrPreview(SubmissionDraft draft, int maxDurationMs)
        {
            var start = draft.StartMs;
            if (start == null)
            {
                throw new SegmentValidationError($"{draft.Segment.DisplayName()} start is required");
            }

            AssertTimestampBounds(start.Value);

            if (draft.Segment == SegmentType.Credits && start != 0 && start < MinDurationMs)
            {
                throw new SegmentValidationError($"Credits start must be 0 or at least {MinDurationMs / 1000}s");
            }

            var end = draft.EndMs;
            if (end == null && draft.Segment == SegmentType.Credits && draft.VideoDurationMs != null)
            {
                end = draft.VideoDurationMs;
            }

            if (end != null)
            {
                AssertTimestampBounds(end.Value);
                if (end <= start)
                {
                    throw new SegmentValidationError("End must be greater than start");
                }

                var duration = end.Value - start.Value;
                if (duration < MinDurationMs || duration > maxDurationMs)
                {
                    throw new SegmentValidationError(
                        $"{draft.Segment.DisplayName()} duration must be between " +
                        $"{MinDurationMs / 1000}s and {maxDurationMs / 1000}s"
                    );
                }
            }

            return new Dictionary<string, object?>
            {
                { "tmdb_id", draft.TmdbId },
                { "type", draft.MediaType },
                { "segment", draft.Segment },
                { "season", draft.Season },
                { "episode", draft.Episode },
                { "start_ms", start },
                { "end_ms", end },
                { "imdb_id", NormalizeImdb(draft.ImdbId) }
            };
        }

        private static void AssertTimestampBounds(int value)
        {
            if (value < 0 || value > MaxTimestampMs)
            {
                throw new SegmentValidationError($"Timestamp must be between 0 and {MaxTimestampMs}");
            }
        }

        private static string? NormalizeImdb(string? raw)
        {
            if (raw == null)
                return null;
            var trimmed = raw.Trim();
            return string.IsNullOrEmpty(trimmed) ? null : trimmed;
        }

        private static bool IsValidImdb(string imdb)
        {
            return ImdbPattern.IsMatch(imdb);
        }
    }
}

File: macos_native/Sources/Segmenter/Services/SegmentValidator.swift
import Foundation

public enum SegmentValidationError: LocalizedError {
    case invalidMessage(String)

    public var errorDescription: String? {
        switch self {
        case .invalidMessage(let msg):
            return msg
        }
    }
}

public final class SegmentValidator {
    public static let minDurationMs: Int = 5_000
    public static let maxTimestampMs: Int = 21_600_000 // 6 hours
    public static let maxIntroDurationMs: Int = 200_000
    public static let maxRecapDurationMs: Int = 1_200_000
    public static let maxCreditsDurationMs: Int = 1_800_000
    public static let maxPreviewDurationMs: Int = 1_800_000

    private static let imdbPattern = try! NSRegularExpression(pattern: "^tt[0-9]{7,8}$")

    public static func makeTheIntroDBSubmissionRequest(draft: SubmissionDraft) throws -> [String: Any] {
        let imdbId = normalizeImdb(draft.imdbId)
        let hasTmdb = draft.tmdbId > 0
        let hasImdb = imdbId != nil && isValidImdb(imdbId!)

        guard hasTmdb || hasImdb else {
            throw SegmentValidationError.invalidMessage("At least TMDB ID or a valid IMDB ID must be provided")
        }

        if draft.mediaType == .movie {
            if draft.season != nil || draft.episode != nil {
                throw SegmentValidationError.invalidMessage("Season and episode must be empty for movie submissions")
            }
        } else if draft.mediaType == .tv {
            guard let season = draft.season, season > 0,
                  let episode = draft.episode, episode > 0 else {
                throw SegmentValidationError.invalidMessage("Season and episode are required for TV submissions")
            }
        }

        let validated = try validateDraftRange(draft)

        var payload: [String: Any] = [
            "type": (validated["type"] as! MediaType).rawValue,
            "segment": (validated["segment"] as! SegmentType).rawValue,
            "start_ms": validated["start_ms"] as! Int
        ]

        if hasTmdb {
            payload["tmdb_id"] = validated["tmdb_id"]
        } else {
            payload["tmdb_id"] = NSNull()
        }

        if let season = validated["season"] as? Int {
            payload["season"] = String(season)
        }
        if let episode = validated["episode"] as? Int {
            payload["episode"] = String(episode)
        }
        if let endMs = validated["end_ms"] as? Int {
            payload["end_ms"] = endMs
        }
        if let imdb = validated["imdb_id"] as? String {
            payload["imdb_id"] = imdb
        }
        if let videoDuration = draft.videoDurationMs {
            payload["video_duration_ms"] = videoDuration
        }

        return payload
    }

    public static func makeIntroDBSubmissionRequest(draft: SubmissionDraft) throws -> [String: Any] {
        guard draft.mediaType == .tv else {
            throw SegmentValidationError.invalidMessage("IntroDB supports TV episodes only")
        }

        guard let imdbId = normalizeImdb(draft.imdbId), isValidImdb(imdbId) else {
            throw SegmentValidationError.invalidMessage("Valid IMDB ID is required for IntroDB uploads")
        }

        guard let season = draft.season, season > 0,
              let episode = draft.episode, episode > 0 else {
            throw SegmentValidationError.invalidMessage("Season and episode are required for IntroDB uploads")
        }

        guard let introdbSegment = toIntroDBSegmentType(draft.segment) else {
            throw SegmentValidationError.invalidMessage("IntroDB does not support \(draft.segment.displayName) uploads")
        }

        let validated = try validateDraftRange(draft)
        guard let endMs = validated["end_ms"] as? Int else {
            throw SegmentValidationError.invalidMessage("IntroDB requires an explicit end timestamp")
        }

        let startMs = validated["start_ms"] as! Int

        return [
            "segment_type": introdbSegment,
            "imdb_id": imdbId,
            "season": season,
            "episode": episode,
            "start_sec": Double(startMs) / 1000.0,
            "end_sec": Double(endMs) / 1000.0,
            "tvdb_id": NSNull(),
            "tmdb_id": draft.tmdbId > 0 ? draft.tmdbId : NSNull()
        ]
    }

    private static func toIntroDBSegmentType(_ segment: SegmentType) -> String? {
        switch segment {
        case .intro: return "intro"
        case .recap: return "recap"
        case .credits: return "outro"
        case .preview: return nil
        }
    }

    private static func validateDraftRange(_ draft: SubmissionDraft) throws -> [String: Any] {
        switch draft.segment {
        case .intro:
            return try validateIntroOrRecap(draft, maxDurationMs: maxIntroDurationMs)
        case .recap:
            return try validateIntroOrRecap(draft, maxDurationMs: maxRecapDurationMs)
        case .credits:
            return try validateCreditsOrPreview(draft, maxDurationMs: maxCreditsDurationMs)
        case .preview:
            return try validateCreditsOrPreview(draft, maxDurationMs: maxPreviewDurationMs)
        }
    }

    private static func validateIntroOrRecap(_ draft: SubmissionDraft, maxDurationMs: Int) throws -> [String: Any] {
        var start = draft.startMs
        if start == nil {
            if draft.segment == .intro {
                start = 0
            } else {
                throw SegmentValidationError.invalidMessage("\(draft.segment.displayName) start is required")
            }
        } else {
            start = max(start!, 0)
        }

        guard let end = draft.endMs else {
            throw SegmentValidationError.invalidMessage("\(draft.segment.displayName) end is required")
        }

        try assertTimestampBounds(start!)
        try assertTimestampBounds(end)

        guard end >= start! else {
            throw SegmentValidationError.invalidMessage("End must be greater than or equal to start")
        }

        let duration = end - start!
        if duration != 0 {
            guard duration >= minDurationMs && duration <= maxDurationMs else {
                throw SegmentValidationError.invalidMessage("\(draft.segment.displayName) duration must be 0 or between \(minDurationMs / 1000)s and \(maxDurationMs / 1000)s")
            }
        }

        return [
            "tmdb_id": draft.tmdbId,
            "type": draft.mediaType,
            "segment": draft.segment,
            "season": draft.season as Any,
            "episode": draft.episode as Any,
            "start_ms": start!,
            "end_ms": end,
            "imdb_id": normalizeImdb(draft.imdbId) as Any
        ]
    }

    private static func validateCreditsOrPreview(_ draft: SubmissionDraft, maxDurationMs: Int) throws -> [String: Any] {
        guard let start = draft.startMs else {
            throw SegmentValidationError.invalidMessage("\(draft.segment.displayName) start is required")
        }

        try assertTimestampBounds(start)

        if draft.segment == .credits && start != 0 && start < minDurationMs {
            throw SegmentValidationError.invalidMessage("Credits start must be 0 or at least \(minDurationMs / 1000)s")
        }

        var end = draft.endMs
        if end == nil && draft.segment == .credits, let videoDuration = draft.videoDurationMs {
            end = videoDuration
        }

        if let explicitEnd = end {
            try assertTimestampBounds(explicitEnd)
            guard explicitEnd > start else {
                throw SegmentValidationError.invalidMessage("End must be greater than start")
            }

            let duration = explicitEnd - start
            guard duration >= minDurationMs && duration <= maxDurationMs else {
                throw SegmentValidationError.invalidMessage("\(draft.segment.displayName) duration must be between \(minDurationMs / 1000)s and \(maxDurationMs / 1000)s")
            }
        }

        return [
            "tmdb_id": draft.tmdbId,
            "type": draft.mediaType,
            "segment": draft.segment,
            "season": draft.season as Any,
            "episode": draft.episode as Any,
            "start_ms": start,
            "end_ms": end as Any,
            "imdb_id": normalizeImdb(draft.imdbId) as Any
        ]
    }

    private static func assertTimestampBounds(_ value: Int) throws {
        guard value >= 0 && value <= maxTimestampMs else {
            throw SegmentValidationError.invalidMessage("Timestamp must be between 0 and \(maxTimestampMs)")
        }
    }

    public static func normalizeImdb(_ raw: String?) -> String? {
        guard let trimmed = raw?.trimmingCharacters(in: .whitespacesAndNewlines), !trimmed.isEmpty else {
            return nil
        }
        return trimmed
    }

    public static func isValidImdb(_ imdb: String) -> Bool {
        let range = NSRange(location: 0, length: imdb.utf16.count)
        return imdbPattern.firstMatch(in: imdb, options: [], range: range) != nil
    }
}
